File: Python/smallest-rotation-with-highest-score.py
# Time:  O(n)
# Space: O(n)

class Solution(object):
    def bestRotation(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        N = len(A)
        change = [1] * N
        for i in xrange(N):
            change[(i-A[i]+1)%N] -= 1
        for i in xrange(1, N):
            change[i] += change[i-1]
        return change.index(max(change))

File: C++/smallest-rotation-with-highest-score.cpp
// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int bestRotation(vector<int>& A) {
        const int N = A.size();
        vector<int> change(N);
        for (int i = 0; i < N; ++i) {
            --change[(i - A[i] + 1 + N) % N];
        }
        for (int i = 1; i < N; ++i) {
            change[i] += change[i - 1] + 1;
        }
        return distance(change.begin(), max_element(change.begin(), change.begin() + N));
    }
};
